src/ddsrt/include/dds/ddsrt/process.h

```diff
@@ -41,6 +44,98 @@ typedef pid_t ddsrt_pid_t;
 
+#ifdef DDSRT_USE_PROCESSCREATION
+#define DDSRT_HAVE_PROCESSCREATION 1
+
+/**
+ * @brief Create new process.
+ *
+ * Creates a new process using the provided executable file. It will have
+ * default priority and scheduling.
+ *
+ * Process arguments are represented by argv, which can be null. If argv is
+ * not null, then the array must be null terminated. The argv array only has
+ * to contain the arguments, the executable filename doesn't have to be in
+ * the first element, which is normally the convention.
+ *
+ * @param[in]   executable     Executable file name.
+ * @param[in]   argv           Arguments array.
+ * @param[out]  pid            ID of the created process.
+ *
+ * @returns A dds_retcode_t indicating success or failure.
+ *
+ * @retval DDS_RETCODE_OK
+ *             Process successfully created.
+ * @retval DDS_RETCODE_BAD_PARAMETER
+ *             Provided file is not executable.
+ * @retval DDS_RETCODE_ERROR
+ *             Process could not be created.
+ */
+DDS_EXPORT dds_retcode_t
+ddsrt_process_create(
+  const char *executable,
+  char *const argv[],
+  ddsrt_pid_t *pid);
+
+/**
+ * @brief Wait for a process to exit.
+ *
+ * When the process (identified by pid) has exited, the status argument will
+ * contain the process exit code.
+ *
+ * When the process is terminated by a call to ddsrt_process_terminate(), then
+ * the return code is uncertain, except that it won't be DDS_RETCODE_TIMEOUT.
+ *
+ * @param[in]   pid            Process ID (PID) of the process to wait for.
+ * @param[in]   timeout        Max timeout to wait for the process to exit.
+ * @param[out]  status         The exit code of the process.
+ *
+ * @returns A dds_retcode_t indicating success or failure.
+ *
+ * @retval DDS_RETCODE_OK
+ *             Process has exited.
+ * @retval DDS_RETCODE_TIMEOUT
+ *             Process is still alive.
+ * @retval DDS_RETCODE_BAD_PARAMETER
+ *             Process unknown.
+ * @retval DDS_RETCODE_ERROR
+ *             Wait failed for an unknown reason.
+ */
+DDS_EXPORT dds_retcode_t
+ddsrt_process_wait_exit(
+  ddsrt_pid_t pid,
+  dds_duration_t timeout,
+  int32_t *status);
+
+/**
+ * @brief Terminate a process.
+ *
+ * This will try to gracefully terminate the process (identified by pid).
+ *
+ * If the process hasn't terminated within the timeout, the process will
+ * be forcefully killed.
+ *
+ * @param[in]   pid            Process ID (PID) of the process to terminate.
+ * @param[in]   timeout        Max timeout to wait for graceful termination.
+ *
+ * @returns A dds_retcode_t indicating success or failure.
+ *
+ * @retval DDS_RETCODE_OK
+ *             Process has gracefully terminated.
+ * @retval DDS_RETCODE_TIMEOUT
+ *             Graceful termination failed. Process has been killed.
+ * @retval DDS_RETCODE_BAD_PARAMETER
+ *             Process unknown.
+ * @retval DDS_RETCODE_ERROR
+ *             Termination failed for an unknown reason.
+ */
+DDS_EXPORT dds_retcode_t
+ddsrt_process_terminate(
```

Comment:
I'd prefer to have a `ddsrt_kill` that accepts a signal that can also be used to check if a process is still alive, as is customary on \*NIX systems. Then define DDSRT_TERM and DDSRT_KILL to SIGTERM and SIGKILL respectively and accept `0` to check if the process is alive. I also don't like that the function goes on to forcefully killing a process. That's something that the caller should decide.

Reply:
Process termination has been refactored. It is still called ddsrt_process_terminate(), but there's no wait, and the caller determines if the termination should be forceful or graceful.

Reply:
I would prefer a separate *ddsrt_proc_term* and *ddsrt_proc_kill*. I'd also like a *ddsrt_proc_exists* to check if a certain process with a given pid is running.